### PR TITLE
Issue 377 start firefox from custom path

### PR DESCRIPTION
**Pull request by dwt**
Pull request for https://github.com/cobrateam/splinter/issues/377.

I have to say that I'm not particularly proud of the unit tests as they just mock out half the universe. I'd love to get some guidance how you would consider testing this.

### CODE REVIEW

**Comment by dwt** [comment]
Ah, damn. I forgot to mention that I'm having quite some trouble getting the full test suite to run on my machine - the Firefox driver specific tests seem to work fine, but I'm getting lots of errors that look quite unrelated to me.

**Comment by douglascamata** [comment]
I will take a look tomorrow at lunch time. Already 00:30am for me. Thanks!

**Comment by douglascamata** [comment]
1. Can you, please, break your code into smaller functions, specially the improved guess for OS X Firefox's path?

2. I would make a normal functional test first for Linux that temporarily removes Firefox from the path and try to create a Browser instance passing that Firefox binary path manually, avoiding mocks as much as possible. Then I would create a mock test for the improved OS X guess, mostly because we can't run this on CI server. With the code broken into smaller functions testing OS X specific guess will be easier, I think. But we still don't have a egg for mocking stuff. For this, IMHO, we can add the `mock` egg, which is the official `unittest` mock solution for Python 3.3 and beyond. What you think, @andrewsmedina?

3. Also, please move `selenium.webdriver.firefox.firefox_binary.FirefoxBinary` import to the top of the Firefox webdriver. 

4. Documentation should describe the behavior when two Firefox binaries are found in OS X

Thank you very much!

**Comment by dwt** [comment]
Hi @douglascamata,

I tried to heed your recommendations, however I encountered some issues doing so.

1. Extracting the guessing method, didn't do too much good, as the constructor WebDriver() already has the unwanted side effect of actually opening the browser, so I still needed the mocks. :(

2. Rewriting the test for the setting of a custom path to extract FF from the path and see that it is then set from the right path is a test that would always fail on anything but linux, as the whole point of this patch is that selenium has quite some autodetection on where to find FF on Linux and Windows - but not on mac. See also: https://github.com/SeleniumHQ/selenium/issues/1128 . So what to do? I'd like to provide a test that I can a) test myself (mac user) and b) works on any platform so other people don't see spurious failures running the test. Also I'd like not to restrict the test to only darwin, so you can run it on the integration server.

3. I haven't yet pulled in mock, as @andrewsmedina hasn't yet said anything - in the meantime
   ```python
from contextlib import contextmanager
@contextmanager
def spy_on(an_object, slot_name, spy):
    original_value = getattr(an_object, slot_name)
    setattr(an_object, slot_name, spy)
    try:
        yield spy
    finally:
        setattr(an_object, slot_name, original_value)
```
isn't too bad I think, and should be quite easy to replace once you decide to pull in mock.

What do you think?

**Comment by dwt** [comment]
Hi friends,

any update on this, did you have time yet to take a look at my changes?

Best regards,
Martin

**Comment by douglascamata** [comment]
@dwt I agree with all your points and read your issue at Selenium (I see did you maybe will have some progress there with a contribution). I would just like to wait for another colleague review, but it seems that @andrewsmedina is very busy. @fsouza, can you do a quick review here for me, please?

**Comment by josemazo** [comment]
Hey guys? How is this new feature going? A soon 0.7.4 release with Firefox path selector will be amazing!

**Comment by dwt** [comment]
I'm currently trying to find the time to port the auto detection part of this patch over to selenium.

I'd love to see this go in to splinter in the meantime as this should be a nice usability boon. As soon as selenium has proper has the auto detection included and released, it could be removed from splinter again, but I guess that will take some time.

**Comment by trojkat** [comment]
Can we skip the auto detection part and merge only the `firefox_binary_path` feature? C'mon guys, let's to this!

**Comment by bmcculley** [comment]
@trojkat it's already possible to set a custom Firefox binary path.

```
from splinter import Browser

binary_path = {'binary_path':'<path/to/binary>'}

with Browser('firefox', **binary_path) as browser:
    browser.visit('http://example.com')
    browser.quit()
```

**Comment by trojkat** [comment]
@bmcculley with this patch yes, but not with current master
https://github.com/cobrateam/splinter/blob/master/splinter/driver/webdriver/firefox.py#L20

**Comment by andrewsmedina** [comment]
@trojkat I will take a time to merge the custom binary path. So @dwt can open another PR with the auto detection.

File: splinter/driver/webdriver/firefox.py [hunk 1]
```diff
@@ -42,9 +45,16 @@ def __init__(self, profile=None, extensions=None, user_agent=None,
         if extensions:
             for extension in extensions:
                 firefox_profile.add_extension(extension)
-
+        
+        firefox_binary = None
+        if firefox_binary_path is not None:
```
Comment (amcorreia) [inline splinter/driver/webdriver/firefox.py]: This should be:
if firefox_binary_path:
More pythonic way...

File: splinter/driver/webdriver/firefox.py [hunk 2]
```diff
@@ -54,3 +64,23 @@ def __init__(self, profile=None, extensions=None, user_agent=None,
...
+    def _guess_firefox_path_if_sensible(self):
+        # Selenium guesses really badly (i.e. not at all) where Firefox on os x is installed.
+        # So let's ask the system instead
+        standard_firefox_path = "/Applications/Firefox.app/Contents/MacOS/firefox-bin"
+        if not sys.platform.startswith('darwin') \
+            or os.path.isfile(standard_firefox_path):
+            return None
+        
+        path = check_output(["mdfind", "kMDItemFSName = Firefox.app"]).decode('utf-8').strip()
+        if path == '': return None
```
Comment (amcorreia) [inline splinter/driver/webdriver/firefox.py]: if not path:
  return None

**Comment by andrewsmedina** [comment]
I will close this PR due to inactivity. I someone want to continue this work in the future, feel free to open another PR.

**Pull request closed without merging**